### PR TITLE
Add support for multiplication

### PR DESCRIPTION
**Pull request by tohrnii**
Addressing #12.
Adds multiplication to entire pipeline.

- [x] scanning
- [x] parsing
- [x] IR
- [x] codgen (doesn't include unit test for codegen)

### CODE REVIEW

**Review by grjte — CHANGES_REQUESTED** [review]
Thanks @tohrnii, it looks good! My comments are mostly related to doing more rigorous testing - I think this is a good time for us to start addressing edge cases more systematically.

One other general nit: when a `clippy` or `rustfmt` error is introduce in a particular PR, I would prefer to rebase the fix. A "fix" commit is fine if the error was introduced in a different PR, but in a case like this it's a simple rebase, and it will keep the git history cleaner.

File: parser/src/lexer/tests.rs [hunk 1]
```diff
@@ -45,11 +45,16 @@ fn keywords_with_identifiers() {
 
 #[test]
 fn multi_arithmetic_ops() {
-    let source = "enf clk' - clk - 1 = 0";
+    let source = "enf clk' * clk - clk' - clk - 1 = 0";
```
[the remark this answers is not in the document]
Comment (tohrnii) [reply parser/src/lexer/tests.rs]: I've renamed these to `chained_<op>_name` but I'm not sure if there's much utility in doing extensive testing for lexer for chained operations since these should only cause error at the parser level. Maybe all of lexer tests should be structured differently than I've done in the updated PR. Please let me know your thoughts on this.

File: parser/src/parser/grammar.lalrpop [hunk 2]
```diff
@@ -87,8 +87,13 @@ TransitionConstraint: TransitionConstraint = {
...
 Expr: Expr = {
-    <lexpr: Expr> "+" <rexpr: NextExpr> => Expr::Add(Box::new(lexpr), Box::new(rexpr)),
-    <lexpr: Expr> "-" <rexpr: NextExpr> => Expr::Subtract(Box::new(lexpr), Box::new(rexpr)),
+    <lexpr: Expr> "+" <rexpr: MulExpr> => Expr::Add(Box::new(lexpr), Box::new(rexpr)),
+    <lexpr: Expr> "-" <rexpr: MulExpr> => Expr::Subtract(Box::new(lexpr), Box::new(rexpr)),
+    MulExpr
+}
+
+MulExpr: Expr = {
+    <lexpr: NextExpr> "*" <rexpr: NextExpr> => Expr::Multiply(Box::new(lexpr), Box::new(rexpr)),
```
[the remark this answers is not in the document]
Comment (tohrnii) [reply parser/src/parser/grammar.lalrpop]: Thank you. Fixed.

File: parser/src/parser/tests/expressions.rs [hunk 3]
```diff
@@ -5,6 +5,30 @@ use super::{
 
 // SECTIONS
 // ================================================================================================
+#[test]
+fn multiplication() {
+    // the operation must be put into a source section, or parsing will fail
+    let source = "
+    transition_constraints:
+        enf clk' * clk - 1 = 0";
```
[the remark this answers is not in the document]
Comment (tohrnii) [reply parser/src/parser/tests/expressions.rs]: Please let me know if it's ok now.

File: codegen-winter/src/air/transition_constraints.rs [hunk 4]
```diff
@@ -51,43 +51,38 @@ impl Codegen for NodeIndex {
...
             Operation::Add(l_idx, r_idx) => {
                 let lhs = l_idx.to_string(graph);
-                strings.push(lhs);
 
                 // output Add followed by Neg as "-"
-                let r_idx = if let Operation::Neg(n_idx) = graph.node(r_idx).op() {
-                    strings.push(String::from("-"));
-                    n_idx
+                let rhs = if let Operation::Neg(n_idx) = graph.node(r_idx).op() {
+                    format!("-{}", n_idx.to_string(graph))
```
Comment (grjte) [inline codegen-winter/src/air/transition_constraints.rs:71]: We actually need to put parentheses around the expression being negated. There was an error here before.

The most basic fix would be this:
```
format!("- ({})", n_idx.to_string(graph))
```

It would be nice to do something more sophisticated and only put the parenthesis if expression represented by `n_idx` is actually an operation. In that case, this part should be pulled into its own helper method

File: codegen-winter/src/air/transition_constraints.rs [hunk 5]
```diff
@@ -51,43 +51,38 @@ impl Codegen for NodeIndex {
...
                 };
-
+                format!("{}{}", lhs, rhs)
+            }
+            Operation::Mul(l_idx, r_idx) => {
+                let lhs = l_idx.to_string(graph);
                 let rhs = r_idx.to_string(graph);
-
-                strings.push(rhs);
+                format!("{}*{}", lhs, rhs)
```
Comment (grjte) [inline codegen-winter/src/air/transition_constraints.rs:80]: Similar comment to the one for negation above - We need to have parentheses around the sub-expressions if they are addition expressions. It would be nice to have a shared helper for this and the negation case above

File: parser/src/lexer/tests.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (grjte) [reply parser/src/lexer/tests.rs]: I think you're right about most of the complex tests being at the parser level. We can audit the tests later to assess which edge cases we're missing and which tests are redundant. For now, I think what you have is good, and I like the restructure.

**Review by grjte — CHANGES_REQUESTED** [review]
Thanks @tohrnii ! The changes all look great. Unfortunately I think there's a bug that I didn't catch in the last pass (it was actually there previously). I left 2 comments inline. Once those are fixed, we can merge this. If you want, feel free to do the simple naive solution now so we can get this merged and open an issue for a more elegant solution

File: codegen-winter/src/air/transition_constraints.rs (hunk 4 above)
Comment (grjte) [reply codegen-winter/src/air/transition_constraints.rs:71]: It might make sense to do the more sophisticated fix in a future PR so we can merge this one.

File: codegen-winter/src/air/transition_constraints.rs (hunk 5 above)
Comment (grjte) [reply codegen-winter/src/air/transition_constraints.rs:80]: It might make sense to do the more sophisticated fix in a future PR so we can merge this one and just do the following for now:

```
format!("({}) * ({})", lhs, rhs)
```

File: codegen-winter/src/air/transition_constraints.rs [hunk 6]
```diff
@@ -50,44 +50,40 @@ impl Codegen for NodeIndex {
...
             }
             Operation::MainTraceNextRow(col_idx) => {
-                strings.push(format!("next[{}]", col_idx));
+                format!("next[{}]", col_idx)
             }
             Operation::Neg(idx) => {
-                strings.push(String::from("-"));
                 let str = idx.to_string(graph);
-                strings.push(str);
+                format!("-{}", str)
```
Comment (grjte) [inline codegen-winter/src/air/transition_constraints.rs:65]: Ah sorry, I missed one! We need to do the same thing with this one too.

**Review by grjte — CHANGES_REQUESTED** [review]
Sorry @tohrnii there's one more instance of the same issue...

**Comment by grjte** [comment]
@tohrnii can you also open an issue for implementing a more elegant solution to nested expressions during code generation?

**Review by grjte — APPROVED** [review]
It looks great now. Thanks @tohrnii!
